src/context.rs

```diff
@@ -120,6 +131,31 @@ impl ApplicationContextData {
 
         self.config.update_connection_state(state);
     }
+
+    /// Add a new event listener.
+    fn add_event_listener<T>(&mut self, listener: T)
+    where
+        T: 'static + ApplicationEventListener + Send,
+    {
+        self.event_listeners.push(Box::new(listener));
+    }
+
+    /// Add all given event listeners.
+    fn add_event_listeners(
+        &mut self,
+        mut listeners: Vec<Box<dyn ApplicationEventListener + Send>>,
+    ) {
+        for listener in self.event_listeners.drain(..) {
+            listeners.push(listener);
+        }
+
+        std::mem::replace(&mut self.event_listeners, listeners);
```

Comment:
It's because of the use case. When you look at the `set_scanning` and `set_connection_state` methods, you'll see that these methods take all listeners using the `take_event_listeners` and then return them back using this method.

The assumption is that it is quite unlikely that there will be any new listeners in `self.event_listeners` during that short period. Therefore, it is much more effective to put back the original vector and append (most likely empty) `self.event_listeners` to it. This way we avoid unnecessary allocations on every event dispatch.